### DISCUSSION TITLE
"Allow only safe HTML" channel setting appears to disable a range of tags that are not obviously unsafe

### DISCUSSION THREAD

**Discussion by PaulBaileyAtChan** [ExpressionEngine/ExpressionEngine#2939] [Feature Requests]
The "Allow only safe HTML" option within channel settings disables some tags (presumably those deemed "unsafe"), by converting their angle brackets to HTML entities. The set of HTML tags disabled in this way seems surprisingly large, and not obviously connected to them being unsafe.

For example, the following tags appear to be disabled (converted to HTML entities) when "Allow only safe HTML" is turned on:

```
<h1> (but not, confusingly, <h2> or <h3>)
<ul>
<ol>
<li>
<div>
```

It isn't clear that these tags are unsafe.

This has been tested in EE5.4.1, and a recent discussion on the official EE Slack channel suggests that the behaviour is the same in EE6.

One aspect of this issue is that neither the EE documentation, nor the EE control panel, properly describes what this option does, and which tags are affected, so that a user can make an informed choice about whether it should be turned on — so its behaviour is impenetrable.

It is suggested that a review should be made of which tags are affected by this option, and that this option should then be more fully documented, in both the EE docs and the EE control panel.

**Comment by jcogs-design** [comment]
This appears to be reporting actual behaviour - the translations of HTML into 'safe HTML' appear to take place within the definition of the Typography class ( in **legacy/libraries/Typography.php** ) - the translation rules appear to be defined between lines 154 and 204.  In summary they appear to allow the following tags: 
- `b`,`blockquote`,`cite`,`code`,`del`,`em`,`i`,`ins`,`mark`,`pre`,`span`,`strong`,`sub`,`sup`
- `h2`, `h3`, `h4`, `h5`, `h6`
which is consistent with the findings of the intial poster.

Quite why `h1` and `p` are not included (for example) is a mystery - the code includes no comments explaining what the basis for these choices is, and [the documentation of this feature](https://docs.expressionengine.com/v5/control-panel/channels.html#settings-tab) (as noted in initial posting) has been missing some text at least since V5.

It is also unclear what the purpose of the feature is. SafeHTML appears to be something that was important about 2007, where a number of initiatives were launched to define HTML that was immune from XSS attacks, but other protections have since been implemented, so possibly it is now a redundant feature.  

If this function is to be implemented, a better list of 'safe' tags is probably required. Due to the largely missing information from the work done in 2007-2010 on this, such lists are not easy to find.  However [this stackExchange posting](https://meta.stackexchange.com/a/135909/770940) appears to offer a fairly comprehensive white list (that includes `h1` and `p`!!!) along with some explanations for the thinking involved.  Another similar list (but in less useful format) can be found [here](https://docs.jboss.org/hibernate/validator/4.2/api/org/hibernate/validator/constraints/SafeHtml.WhiteListType.html).

I'm not sure how useful this function is, however if it is to be retained there are two recommendations:
- The function is redefined to allow a more coherent set of tags along lines of whitelist suggestions linked to in this post
- The documentation of the feature is corrected / completed, including a full list of what HTML tags are allowed through as 'safe'.
HTH
